File: Harvest/ViewModels/FieldViewModel.swift
import Foundation
import Observation

@Observable
final class FieldViewModel {
    var available: [Community] = []
    var joinedIds: Set<String> = []
    var isLoading = false
    var error: String?

    private let service = CommunityService()

    func load(userId: String) async {
        isLoading = true
        defer { isLoading = false }
        do {
            async let avail = service.availableCommunities(userId: userId)
            async let joined = service.joinedCommunityIds(userId: userId)
            self.available = try await avail
            self.joinedIds = try await joined
        } catch {
            self.error = error.localizedDescription
        }
    }

    func toggleJoin(_ community: Community, userId: String) async {
        do {
            if joinedIds.contains(community.id) {
                try await service.leave(communityId: community.id, userId: userId)
                joinedIds.remove(community.id)
            } else {
                try await service.join(communityId: community.id, userId: userId)
                joinedIds.insert(community.id)
            }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func isJoined(_ community: Community) -> Bool { joinedIds.contains(community.id) }
}

File: android/app/src/main/java/com/harvestglass/harvest/ui/field/FieldViewModel.kt
package com.harvestglass.harvest.ui.field

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.harvestglass.harvest.data.model.Community
import com.harvestglass.harvest.data.service.CommunityService
import com.harvestglass.harvest.util.userMessage
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class FieldUiState(
    val available: List<Community> = emptyList(),
    val joinedIds: Set<String> = emptySet(),
    val isLoading: Boolean = false,
    val error: String? = null
) {
    fun isJoined(community: Community): Boolean = joinedIds.contains(community.id)
}

/** Mirrors Harvest/ViewModels/FieldViewModel.swift. */
@HiltViewModel
class FieldViewModel @Inject constructor(
    private val service: CommunityService
) : ViewModel() {

    private val _state = MutableStateFlow(FieldUiState())
    val state: StateFlow<FieldUiState> = _state.asStateFlow()

    fun load(userId: String) = viewModelScope.launch {
        _state.update { it.copy(isLoading = true) }
        try {
            // iOS runs both concurrently via `async let`.
            coroutineScope {
                val availableDeferred = async { service.availableCommunities(userId) }
                val joinedDeferred = async { service.joinedCommunityIds(userId) }
                val available = availableDeferred.await()
                val joined = joinedDeferred.await()
                _state.update { it.copy(available = available, joinedIds = joined) }
            }
        } catch (e: Exception) {
            _state.update { it.copy(error = e.userMessage()) }
        } finally {
            // Mirrors the Swift `defer` — clears on failure too.
            _state.update { it.copy(isLoading = false) }
        }
    }

    fun toggleJoin(community: Community, userId: String) = viewModelScope.launch {
        try {
            if (_state.value.joinedIds.contains(community.id)) {
                service.leave(community.id, userId)
                _state.update { it.copy(joinedIds = it.joinedIds - community.id) }
            } else {
                service.join(community.id, userId)
                _state.update { it.copy(joinedIds = it.joinedIds + community.id) }
            }
        } catch (e: Exception) {
            _state.update { it.copy(error = e.userMessage()) }
        }
    }
}
